text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,801 @@
+
+
+### Unicode Identifiers
+
+Rust allows Unicode identifiers but our character set is restricted to ASCII alphanumerics, and `_`. In order to transcode the former to the latter, we use the same approach as Swift, which is: encode all non-ascii identifiers via [Punycode][punycode], a standardized and efficient encoding that keeps encoded strings in a rather human-readable format. So for example, the string
+
+```
+"Gödel, Escher, Bach"
+```
+
+is encoded as
+
+```
+"Gdel, Escher, Bach-d3b"
+```
+
+which, as opposed to something like _Base64_, still gives a pretty good idea of what the original string looked like.
+
+Each component of a name, i.e. anything that starts with the number of bytes to read in the examples above, is encoded individually. Components encoded this way are augmented with a `u` suffix so that demanglers know that the identifier needs further decoding. As an example, the function:
+
+```rust
+mod gödel {
+  mod escher {
+    fn bach() {}
+  }
+}
+```
+
+would be mangled as:
+
+```
+_RN15mycrate_4a3b56d8gdel_Fqau6escher4bachVE`
+                    <-------->
+                Unicode component
+```
+
+### Compression/Substitution
+
+The length of symbol names has an influence on how much work compiler, linker, and loader have to perform. The shorter the names, the better. At the same time, Rust's generics can lead to rather long names (which are often not visible in the code because of type inference and `impl Trait`). For example, the return type of the following function:
+
+```rust
+fn quux(s: Vec<u32>) -> impl Iterator<Item=(u32, usize)> {
+    s.into_iter()
+     .map(|x| x+1)
+     .filter(|&x| x > 10)
+     .zip(0..)
+     .chain(iter::once((0, 0)))
+}
+```
+
+is
+
+```rust
+std::iter::Chain<
+  std::iter::Zip<
+    std::iter::Filter<
+      std::iter::Map<
+        std::vec::IntoIter<u32>,
+        [closure@src/main.rs:16:11: 16:18]>,
+      [closure@src/main.rs:17:14: 17:25]>,
+    std::ops::RangeFrom<usize>>,
+  std::iter::Once<(u32, usize)>>
+```
+
+It would make for a long symbol name if this types is used (maybe repeatedly) as a generic argument somewhere. C++ has the same problem with its templates; which is why the Itanium mangling introduces the concept of compression. If a component of a definition occurs more than once, it will not be repeated and instead be emitted as a substitution marker that allows to reconstruct which component it refers to. The scheme proposed here will use the same approach.
```

Comment:
```suggestion
It would make for a long symbol name if this type is used (maybe repeatedly) as a generic argument somewhere. C++ has the same problem with its templates; which is why the Itanium mangling introduces the concept of compression. If a component of a definition occurs more than once, it will not be repeated and instead be emitted as a substitution marker that allows to reconstruct which component it refers to. The scheme proposed here will use the same approach.
```